text/0000-with-bounds-in-pi-types.md

```diff
@@ -0,0 +1,685 @@
+checking if the parameters satisfy the trait bounds). The caller's bounds must
+imply the invoked functions' bounds:
+
+### Transitivity of bounds
+
+We require a bound of a function to imply the bounds of the functions it calls,
+through a simple reductive, unification algorithm. In particular, this means
+that a statement is reduced by some specified rules (see below), that ensures
+termination. A statement implies another statement, if the set of statements it
+reduces to is a superset of the other statement's reduction set.
+
+The compiler would enforce that if `f` calls `g`, `unify(bound(g)) ⊆
+unify(bound(f))` (by structural equality):
+
+    ExpandBooleanAnd:
+      P ∧ Q
+      ─────
+      P
+      Q
+
+This simply means that `a ∧ b` means `a` and `b`.
+
+    SubstituteEquality:
+      P(a)
+      a = b
+      ─────
+      P(b)
+
+This is an important inference rule, when doing unification. This means that
+you can substitute all `a` for all free `b`s, if `a = b`.
+
+    DoubleNegation:
+      ¬¬x
+      ───
+      x
+
+This rule is simply stating that double negation is identity, that is, `!!a`
+means that `a` is true.
+
+These rules are "eliminatory" (recursing downwards the tree and decreasing the
+structure), and thus it is possible to check, in this language, that `a ⇒ b`
+relatively quickly (`O(n)`). For a proof of see the section below.
+
+More rules can be added in the future. It is however important to preserve the
+"sequential property" (that is, each step is a reduction, not an expansion),
+allowing one to check the implication in linear time.
+
+This is done under type unification. Thus, we only need to check the bounds at
+the top level.
+
+#### Decidability of this rule set
+
+One can show this by considering each case:
+
+1. `ExpandBooleanAnd` eliminates `{P ∧ Q} ⊢ {P, Q}`. The right hand side's
+   depth is `max(dep(P), dep(Q))`, which is smaller than the original,
+   `max(dep(P), dep(Q)) + 1`
+2. `SubstituteEquality` eliminates `{a = b, P} ⊢ {P[b ← a]}`, which is an
+   elimination, since `dep(P) + 1 > dep(P[b ← a]) = dep(P)`.
+3. `DoubleNegation` eliminates `{¬¬x} ⊢ {x}`, which is an elimination, since
+   `dep(x) + 2 > dep(x)`.
+
+In fact, this set of rule is strictly reductive (like equality-based unification).
+
+#### An example
+
+We will quickly give an example of a possible proof. Say we want to show that
+`(x = b) ∧ ¬¬(x < a) ⇒ b < a`. Starting with the left hand side, we can sequentially
+prove this, by simple unification (which already exists in the Rust type
+checker):
+
+    (x = b) ∧ ¬¬(x < a)
+    ∴ x = b      (ExpandBooleanAnd)
+      ¬¬(x < a)
+    ∴ ¬¬(b < a)  (SubstituteEquality)
+    ∴ b < a      (DoubleNegation)
+      ¯¯¯¯¯
+
+### Contradictive or unsatisfiable bounds
+
+Contradictive or unsatisfiable bounds (like `a < b, b < a`) cannot be detected,
+since such a thing would be undecidable.
+
+These bounds don't break anything, they are simply malformed and unreachable.
```

Comment:
Maybe this is the same question as my last comment, but does this also mean "it's not a compiler error to declare such an item, that item would simply never get used because no call site could ever satisfy its bounds"?

Also, what happens if I create an item `with a<b, b<a` then create a pathological Ord impl where a<b && b<a is actually possible? Does it then get invoked "normally"?